src/couch_jobs/test/couch_jobs_tests.erl

```diff
@@ -0,0 +1,607 @@
+    ?_test(begin
+        Accept = fun() -> exit(couch_jobs:accept(T)) end,
+        WaitAccept = fun(Ref) ->
+            receive
+                {'DOWN', Ref, _, _, Res} -> Res
+            after
+                500 -> timeout
+            end
+        end,
+        {_, Ref1} = spawn_monitor(Accept),
+        ok = couch_jobs:add(?TX, T, J1, #{}),
+        ?assertMatch({ok, #{id := J1}, #{}}, WaitAccept(Ref1)),
+        {_, Ref2} = spawn_monitor(Accept),
+        ?assertEqual(timeout, WaitAccept(Ref2)),
+        ok = couch_jobs:add(?TX, T, J2, #{}),
+        ?assertMatch({ok, #{id := J2}, #{}}, WaitAccept(Ref2))
+    end).
+
+
+job_processor_update(#{t1 := T, j1 := J}) ->
+    ?_test(begin
+        ok = couch_jobs:add(?TX, T, J, #{}),
+        {ok, Job, #{}} = couch_jobs:accept(T),
+
+        % Use proper transactions in a few places here instead of passing in
+        % ?TX This is mostly to increase code coverage
+
+        ?assertMatch({ok, #{job := true}}, fabric2_fdb:transactional(fun(Tx) ->
+            couch_jobs:update(Tx, Job, #{<<"x">> => 1})
+        end)),
+
+        ?assertMatch(#{data := #{<<"x">> := 1}, state := running},
+            get_job(T, J)),
+
+        ?assertMatch({ok, #{job := true}}, fabric2_fdb:transactional(fun(Tx) ->
+            couch_jobs:update(Tx, Job)
+        end)),
+
+        ?assertMatch(#{data := #{<<"x">> := 1}, state := running},
+            get_job(T, J)),
+
+        ?assertMatch({ok, #{job := true}}, fabric2_fdb:transactional(fun(Tx) ->
+            couch_jobs:update(Tx, Job, #{<<"x">> => 2})
+        end)),
+
+        % check json validation for bad data in update
+        ?assertMatch({error, {json_encoding_error, _}},
+            fabric2_fdb:transactional(fun(Tx) ->
+                couch_jobs:update(Tx, Job, #{1 => 1})
+            end)),
+
+        ?assertMatch(#{data := #{<<"x">> := 2}, state := running},
+            get_job(T, J)),
+
+        % Finish may update the data as well
+        ?assertEqual(ok, couch_jobs:finish(?TX, Job, #{<<"x">> => 3})),
+        ?assertMatch(#{data := #{<<"x">> := 3}, state := finished},
+            get_job(T, J))
+    end).
+
+
+resubmit_enqueues_job(#{t1 := T, j1 := J}) ->
+    ?_test(begin
+        ok = couch_jobs:add(?TX, T, J, #{}),
+        {ok, Job1, #{}} = couch_jobs:accept(T),
+        ?assertMatch({ok, _}, couch_jobs:resubmit(?TX, Job1, 6)),
+        ?assertEqual(ok, couch_jobs:finish(?TX, Job1)),
+        ?assertMatch(#{state := pending, stime := 6}, get_job(T, J)),
+        {ok, Job2, #{}} = couch_jobs:accept(T),
+        ?assertEqual(ok, couch_jobs:finish(?TX, Job2)),
+        ?assertMatch(#{state := finished}, get_job(T, J))
+    end).
+
+
+add_with_scheduled_time(#{t1 := T, j1 := J1, j2 := J2}) ->
+    ?_test(begin
+        ?assertEqual(ok, couch_jobs:add(?TX, T, J1, #{}, 5)),
+        ?assertEqual(ok, couch_jobs:add(?TX, T, J2, #{}, 3)),
+        ?assertMatch({ok, #{id := J}, _} when J =:= J1 orelse J =:= J2,
```

Comment:
Should probably have a test that returns nothing when our max_sched_time is less than the jobs submitted start time.

Reply:
Agreed